subprojects/language-jvm/src/main/java/org/gradle/api/tasks/compile/AbstractCompile.java

```diff
@@ -57,22 +64,41 @@ public void setClasspath(FileCollection configuration) {
         this.classpath = configuration;
     }
 
+    /**
+     * Returns the directory property that represents the directory to generate the {@code .class} files into.
+     *
+     * @return The destination directory property.
+     * @since 6.1
+     */
+    @Incubating
+    @OutputDirectory
+    public DirectoryProperty getDestinationDirectory() {
+        return destinationDir;
+    }
+
     /**
      * Returns the directory to generate the {@code .class} files into.
      *
      * @return The destination directory.
+     * @deprecated use {@link #getDestinationDirectory()}
      */
-    @OutputDirectory
+    @Deprecated
+    @ReplacedBy("destinationDirectory")
     public File getDestinationDir() {
-        return destinationDir.getOrNull();
+        // DeprecationLogger.nagUserOfReplacedMethod("getDestinationDir", "getDestinationDirectory");
```

Comment:
This should have deprecation nag, if it's deprecated.

Reply:
The Kotlin plugin will trigger this. Will discuss with them as part of the separate deprecation PR for 6.2